### PR TITLE
Changed procedure annotation

### PR DESCRIPTION
**Pull request by OliviaYtterbrink**
New annotation takes mode as a parameter, which can be one of `READ`, `WRITE`, `SCHEMA`, or `DBMS`.
Deprecates `@PerformsWrites`

changelog: Added mode to Procedure annotation, deprecates PerformsWrites

### CODE REVIEW

**Comment by jexp** [comment]
yay! Thanks a lot!

**Comment by lutovich** [comment]
This is a really nice change! I still think though that `Procedure.Mode = DatabaseManagementSystem` does not make much sense. Maybe we should use `DATABASE_MANAGEMENT` or `USER_MANAGEMENT` instead of `DBMS`.

**Comment by davidegrohmann** [comment]
What happens if someone mix the old annotation with the new ones in an inconsistent way? E.g, something like:

```
 @PerformsWrites
 @Procedure( mode = READ )
```

Can we add tests that ensure that we fail in such cases and we do not accept the procedure?

**Comment by OliviaYtterbrink** [comment]
Good catch, added tests.
`@PerformsWrites` with `mode = READ` or `mode = WRITE` results in a write procedure.
If it is combined with the other modes, it will generate an error.

**Comment by davidegrohmann** [comment]
@OliviaYtterbrink isn't it better to fail in the case `@PerformsWrites` and `mode = READ` instead trying to guess what the user wanted?

**Comment by fickludd** [comment]
@davidegrohmann yes, but we run into problems with backwards compatibility.
`@PerformsWrites @Procedure`
Should be a valid annotation, and this would yield such a collision. On the plus side, mode = READ is a subset of mode = WRITE, so the conflict is not so bad. 

To solve the explicit conflict, we could create a special value `DEFAULT_READ` that could be undocumented, set as default, and be the only allowable combination with `@PerformsWrites`. Then we could make `@PerformsWrites @Procedure( mode = "READ" )` fail.

**Comment by davidegrohmann** [comment]
@fickludd The conflict might be very bad if the user intended the procedures to be read only and forgot to delete the old `PerformsWrites` annotation.

Moreover I am not following what the problem is, the annotation `@Procedure` is old by the param `mode` is brand new, so it should be possible to fail in case of that conflict. No?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java [hunk 1]
```diff
@@ -1103,6 +1103,12 @@ public void uniqueIndexDrop( IndexDescriptor descriptor ) throws DropIndexFailur
         schemaWrite().uniqueIndexDrop( statement, descriptor );
     }
 
+    @Override
+    public RawIterator<Object[], ProcedureException> procedureCallSchema( ProcedureName name, Object[] input ) throws ProcedureException
+    {
+        return callProcedure( name, input, AccessMode.Static.FULL );
```
Comment (henriknyman) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:1109]: Maybe now that we have procedureCallSchema we can lower the access mode of procedureCallWrite from FULL to WRITE (see the TODO in this file)

**Comment by henriknyman** [comment]
We will go with what @davidegrohmann suggested and fail in the case of mixing old and new annotations, but it will be added on by this separate PR: https://github.com/neo4j/neo4j/pull/7488

**Pull request merged**